vscode/src/jsonrpc/agent-protocol.ts

```diff
@@ -522,6 +524,9 @@ export interface ClientCapabilities {
     // convenient for clients that forward the string directly to an underlying
     // webview container.
     webviewMessages?: 'object-encoded' | 'string-encoded' | undefined | null
+    // when a file path is provided, the agent will attempt to deserialize extension state
+    // from the path and append all global state (such as chat interactions) to the path
+    persistencePath?: string | undefined | null
```

Comment:
Capabilities are conventionally enums. I propose we do this:

```suggestion
    // How to deal with vscode.ExtensionContext.globalState.
    // - Stateless: the state does not persist between agent processes. This means the client is responsible for features like managing chat history.
    // - Server managed: the server reads and writes the state without informing the client. The client can optionally customize the file path of the JSON config via `ClientInfo.globalStatePath: string`
    // - Client managed: not implemented yet. When implemented, clients will be able to implement a JSON-RPC request to handle the saving of the client state. This is needed to safely share state between concurrent agent processes (assuming there is one IDE client process managing multiple agent processes).
    globalState?: "stateless" | "server-managed" | "client-managed" | undefined null
```

Reply:
btw, we can use `path.join(codyPaths(), "globalState.json")` as a default location for server-managed state. However, I think we should ideally try to have clients move to client-managed state ASAP to avoid races between concurrent agents.

Reply:
+1, codyPaths is a foot gun because of all clients share a path and it is going to end badly because we have no concurrency control and have given little thought to versioning.

We need product input, or at least being informed, here on what different clients should do--whether state is shared; if so, which state.

Even if the state is not shared between IDEs, we still have concurrency from multiple instances of the same IDE. I think we should copy vscode here, which is using sqlite3 for some consistency. (We don't need sqlite, but we need something with the same row-level consistency model.)

Reply:
After thinking more about this, we maybe just want to bite the bullet and default to `server-managed` for all IDEs using the same sqlite-powered implementation sharing the same `codyPaths()` file path (with option to override it). This is assuming we get some file locking mechanism from sqlite to handle concurrent clients writing to the disk. None of these writes/reads should take a long time to run.

Reply:
I spent quite a long time trying to get better-sqlite3 or node-sqlite3 to work on windows but there are a lot of known issues around building and bundling it (specifically on windows) because it's using bindings to an underlying C library. I think it would be possible to get it working but we'd likely have to bundle the package separately and patch it in.

I backed out all of those changes and swapped to just using [node-localstorage](https://www.npmjs.com/package/node-localstorage) which is marginally better than just writing to files, but I don't think we're going to have a tremendous volume of writes.

However, I'm just now realizing the issue of concurrency. I doubt node-localstorage handles concurrency well. I can test that and then potentially get back to sqlite.

Reply:
I created an [RFC](https://docs.google.com/document/d/1Sy7SBnuM5p3LDYyT0SNQ7_0HKTYeIEv5QjRzeXjZ2Fc/edit?usp=sharing) for this